### ISSUE TITLE
Mapped Types Assumes readonly For Index Signatures

### ISSUE THREAD

**Issue by CatGuardian** [microsoft/TypeScript#30261]
**TypeScript Version:**  3.3.1

**Search Terms:**

**Code**

```ts
export enum OpenTokVideoCodec {
  VP8 = 'VP8',
  H264 = 'H264',
}

class MyObject {
    constructor() {
    }
}

export type MyObjectsByCodec = {
  [K in keyof typeof OpenTokVideoCodec]?: MyObject;
};

const test: MyObjectsByCodec = {};

test['VP8'] = new MyObject(); // This reports that VP8 property is readonly






type VideoCodecKeys = keyof typeof OpenTokVideoCodec;

export type MyObjectsByCodecAlternate = {
  [K in VideoCodecKeys]?: MyObject;
};

const testAlternate: MyObjectsByCodecAlternate = {};

testAlternate['VP8'] = new MyObject(); // This one reports that it isn't readonly and is ok





for (const codec in testAlternate) {
    if (testAlternate.hasOwnProperty(codec)) {

        // When 'No Implicit Any' is on, then this flags as an error saying 'Element implicitly has an 'any' type because type 'MyObjectsByCodecAlternate' has no index signature.' But this does have an index signature.
        // But it goes away if you update the MyObjectsByCodecAlternate index signature to be this: `[K in keyof typeof OpenTokVideoCodec]?: MyObject;` 
        const myObject = testAlternate[codec];

    }
}
```

**Expected behavior:**
I expect to not get this error message: `Cannot assign to 'VP8' because it is a read-only property.` on the line `test['VP8'] = new MyObject();`

I also don't expect to get `Element implicitly has an 'any' type because type 'MyObjectsByCodecAlternate' has no index signature.` Error message on the `const myObject = testAlternate[codec];` Line.

**Actual behavior:**
I get the following error message: `Cannot assign to 'VP8' because it is a read-only property.` on the line `test['VP8'] = new MyObject();` 

**Playground Link:** 

Don't forget to turn 'noImplicitAny' option on. But whether it is on or not only affects the second bug.

http://www.typescriptlang.org/play/index.html#src=export%20enum%20OpenTokVideoCodec%20%7B%0D%0A%20%20VP8%20%3D%20'VP8'%2C%0D%0A%20%20H264%20%3D%20'H264'%2C%0D%0A%7D%0D%0A%0D%0Aclass%20MyObject%20%7B%0D%0A%20%20%20%20constructor()%20%7B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aexport%20type%20MyObjectsByCodec%20%3D%20%7B%0D%0A%20%20%5BK%20in%20keyof%20typeof%20OpenTokVideoCodec%5D%3F%3A%20MyObject%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20test%3A%20MyObjectsByCodec%20%3D%20%7B%7D%3B%0D%0A%0D%0Atest%5B'VP8'%5D%20%3D%20new%20MyObject()%3B%20%2F%2F%20This%20reports%20that%20VP8%20property%20is%20readonly%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Atype%20VideoCodecKeys%20%3D%20keyof%20typeof%20OpenTokVideoCodec%3B%0D%0A%0D%0Aexport%20type%20MyObjectsByCodecAlternate%20%3D%20%7B%0D%0A%20%20%5BK%20in%20VideoCodecKeys%5D%3F%3A%20MyObject%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20testAlternate%3A%20MyObjectsByCodecAlternate%20%3D%20%7B%7D%3B%0D%0A%0D%0AtestAlternate%5B'VP8'%5D%20%3D%20new%20MyObject()%3B%20%2F%2F%20This%20one%20reports%20that%20it%20isn't%20readonly%20and%20is%20ok%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Afor%20(const%20codec%20in%20testAlternate)%20%7B%0D%0A%20%20%20%20if%20(testAlternate.hasOwnProperty(codec))%20%7B%0D%0A%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20When%20'No%20Implicit%20Any'%20is%20on%2C%20then%20this%20flags%20as%20an%20error%20saying%20'Element%20implicitly%20has%20an%20'any'%20type%20because%20type%20'MyObjectsByCodecAlternate'%20has%20no%20index%20signature.'%20But%20this%20does%20have%20an%20index%20signature.%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20But%20it%20goes%20away%20if%20you%20update%20the%20MyObjectsByCodecAlternate%20index%20signature%20to%20be%20this%3A%20%60%5BK%20in%20keyof%20typeof%20OpenTokVideoCodec%5D%3F%3A%20MyObject%3B%60%20%0D%0A%20%20%20%20%20%20%20%20const%20myObject%20%3D%20testAlternate%5Bcodec%5D%3B%0D%0A%0D%0A%20%20%20%20%7D%0D%0A%7D

**Related Issues:**

**Comment by j-oliveras** [comment]
Workaround:
```ts
export type MyObjectsByCodec = {
  -readonly [K in keyof typeof OpenTokVideoCodec]?: MyObject;
};
```
